### PR TITLE
Experiment Initialization failure handling

### PR DESCRIPTION
**Pull request by arjanoop**
Experiment Initialization failure handling

Resolves: [Issue 420](https://github.com/dice-group/gerbil/issues/420)

### CODE REVIEW

File: src/main/java/org/aksw/gerbil/web/response/execution/ExperimentExecutionResponse.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
+package org.aksw.gerbil.web.response.execution;
+
+public class ExperimentExecutionResponse {
```
Comment (MichaelRoeder) [inline src/main/java/org/aksw/gerbil/web/response/execution/ExperimentExecutionResponse.java:3]: Please add javadoc comments.

**Review by MichaelRoeder — CHANGES_REQUESTED** [review]
I think that it is not good practice to return JSON as content type if the message actually does not contain JSON. Please look at the comments in the code for more details.

File: src/main/java/org/aksw/gerbil/exceptions/FaultyConfigurationException.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package org.aksw.gerbil.exceptions;
+
+import org.aksw.gerbil.datatypes.ExperimentTaskConfiguration;
+
+import java.util.List;
+
+public class FaultyConfigurationException extends RuntimeException{
```
[the remark this answers is not in the document]
Comment (arjanoop) [reply src/main/java/org/aksw/gerbil/exceptions/FaultyConfigurationException.java:26]: Added Javadocs

File: src/main/java/org/aksw/gerbil/web/MainController.java [hunk 3]
```diff
@@ -139,12 +145,13 @@ public ModelAndView index() {
...
      * expects a string like {"type":"A2KB","matching": "Mw - weak annotation match"
      * ,"annotator":["A2KB one","A2KB two" ],"dataset":["datasets"]}
-     * 
+     *
      * @param experimentData
      * @return
      */
-    @RequestMapping("/execute")
-    public @ResponseBody String execute(@RequestParam(value = "experimentData") String experimentData) {
+    @GetMapping(value = "/execute", produces = MediaType.APPLICATION_JSON_VALUE)
```
[the remark this answers is not in the document]
Comment (arjanoop) [reply src/main/java/org/aksw/gerbil/web/MainController.java]: Removed

File: src/main/java/org/aksw/gerbil/web/MainController.java [hunk 4]
```diff
@@ -139,12 +145,13 @@ public ModelAndView index() {
...
+    @ResponseBody
```
[the remark this answers is not in the document]
Comment (arjanoop) [reply src/main/java/org/aksw/gerbil/web/MainController.java]: removed

File: src/main/java/org/aksw/gerbil/web/MainController.java [hunk 5]
```diff
@@ -139,12 +145,13 @@ public ModelAndView index() {
...
+    public ResponseEntity<?> execute(@RequestParam(value = "experimentData") String experimentData) {
```
[the remark this answers is not in the document]
Comment (arjanoop) [reply src/main/java/org/aksw/gerbil/web/MainController.java]: Refactored to `ResponseEntity<String>`

File: src/main/java/org/aksw/gerbil/web/MainController.java [hunk 6]
```diff
@@ -180,9 +187,16 @@ public ModelAndView index() {
...
+        try{
+            exp.run();
+        }catch(FaultyConfigurationException e){
+            if(e.getFaultyConfigs().size()==configs.length){
+                return new ResponseEntity<>(new ExperimentExecutionResponse(e.getMessage()), HttpStatus.BAD_REQUEST);
+            }else{
+                return new ResponseEntity<>(new ExperimentExecutionResponse(experimentId,e.getMessage()), HttpStatus.BAD_REQUEST);
+            }
+        }
+        return new ResponseEntity<>(experimentId, HttpStatus.OK);
```
[the remark this answers is not in the document]
Comment (arjanoop) [reply src/main/java/org/aksw/gerbil/web/MainController.java]: Updated the response accordingly 
`ResponseEntity.ok().contentType(MediaType.TEXT_HTML).body(experimentId);`
and similarly for other scenarios with content type `contentType(MediaType.APPLICATION_JSON)`

File: src/main/java/org/aksw/gerbil/web/MainController.java [hunk 7]
```diff
@@ -180,9 +187,16 @@ public ModelAndView index() {
...
-        exp.run();
-
-        return experimentId;
+        try{
+            exp.run();
+        }catch(FaultyConfigurationException e){
+            if(e.getFaultyConfigs().size()==configs.length){
+                return new ResponseEntity<>(new ExperimentExecutionResponse(e.getMessage()), HttpStatus.BAD_REQUEST);
+            }else{
+                return new ResponseEntity<>(new ExperimentExecutionResponse(experimentId,e.getMessage()), HttpStatus.BAD_REQUEST);
```
[the remark this answers is not in the document]
Comment (arjanoop) [reply src/main/java/org/aksw/gerbil/web/MainController.java]: Yes, added `jackson.databind.ObjectMapper` configuration for generating `json`

File: src/main/java/org/aksw/gerbil/web/response/execution/ExperimentExecutionResponse.java (hunk 1 above)
Comment (arjanoop) [reply src/main/java/org/aksw/gerbil/web/response/execution/ExperimentExecutionResponse.java:22]: added javadoc

File: src/main/java/org/aksw/gerbil/web/MainController.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (MichaelRoeder) [reply src/main/java/org/aksw/gerbil/web/MainController.java]: TEXT_HTML is slightly misleading, or? Wouldn't `text/plain` be better?
Apart from that, it sound good to me.

**Review by MichaelRoeder — CHANGES_REQUESTED** [review]
I think `text/plain` would be the better media type when returning the plain experiment ID.

File: src/main/java/org/aksw/gerbil/web/MainController.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (arjanoop) [reply src/main/java/org/aksw/gerbil/web/MainController.java]: Refactored it to `return ResponseEntity.ok().contentType(MediaType.TEXT_PLAIN).body(experimentId);`

**Review by MichaelRoeder — APPROVED** [review]
LGTM

**Pull request merged**